rfcs/0001-syntax-tree-patterns.md

```diff
@@ -0,0 +1,797 @@
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Overview
+
+The following diagram shows the dependencies between the main parts of the proposed solution:
+
+```
+                          Pattern syntax                                     
+                                |                                           
+                                |  parsing / lowering                       
+                                v                                           
+                           PatternTree                                      
+                                ^                                             
+                                |                                             
+                                |                                             
+                          IsMatch trait                                       
+                                |                                             
+                                |                                             
+             +---------------+-----------+---------+                      
+             |               |           |         |                      
+             v               v           v         v                      
+        syntax::ast     rustc::hir      syn       ...               
+```
+
+The pattern syntax described in the previous section is parsed / lowered into the so-called *PatternTree* data structure that represents a valid syntax tree pattern. Matching a *PatternTree* against an actual syntax tree (e.g. rust ast / hir or the syn ast, ...) is done using the *IsMatch* trait.
+
+The *PatternTree* and the *IsMatch* trait are introduced in more detail in the following sections.
+
+## PatternTree
+
+The core data structure of this RFC is the **PatternTree**. 
+
+It's a data structure similar to rust's AST / HIR, but with the following differences:
+
+- The PatternTree doesn't contain parsing information like `Span`s
+- The PatternTree can represent alternatives, sequences and optionals
+
+The code below shows a simplified version of the current PatternTree:
+
+> Note: The current implementation can be found [here](https://github.com/fkohlgrueber/pattern-matching/blob/dfb3bc9fbab69cec7c91e72564a63ebaa2ede638/pattern-match/src/pattern_tree.rs#L50-L96).
+
+
+```
+pub enum Expr {
+    Lit(Alt<Lit>),
+    Array(Seq<Expr>),
+    Block_(Alt<BlockType>),
+    If(Alt<Expr>, Alt<BlockType>, Opt<Expr>),
+    IfLet(
+        Alt<BlockType>,
+        Opt<Expr>,
+    ),
+}
+
+pub enum Lit {
+    Char(Alt<char>),
+    Bool(Alt<bool>),
+    Int(Alt<u128>),
+}
+
+pub enum Stmt {
+    Expr(Alt<Expr>),
+    Semi(Alt<Expr>),
+}
+
+pub enum BlockType {
+    Block(Seq<Stmt>),
+}
+```
+
```

Comment:
Won't we also need to deal with impls, structs, etc? A lot of clippy lints deal with items not expressions.

Of course as a first step this is still amazing :)